lib/promoted/ruby/client.rb

```diff
@@ -31,32 +28,103 @@ def initialize (params={})
+          raise ArgumentError.new("delivery_endpoint is required") if @delivery_endpoint.empty?
+
+          @metrics_endpoint = params[:metrics_endpoint].to_s
+          raise ArgumentError.new("metrics_endpoint is required") if @metrics_endpoint.empty?
 
           @sampler = Sampler.new
         end
         
-        def send_request payload, endpoint=nil
-          endpoint ||= BASE_URL
+        def send_request payload, endpoint, timeout_millis, headers={}
           response = Faraday.post(endpoint) do |req|
-            req.headers['Content-Type'] = 'application/json'
+            req.headers                 = req.headers.merge!(headers) if headers
+            req.headers['Content-Type'] = 'application/json' if req.headers['Content-Type'].blank?
+            req.options.timeout         = timeout_millis / 1000
             req.body                    = payload.to_json
           end
-          response
+  
+          # TODO: Check response code, rescue on ParserError, etc.
+          JSON.parse(response.body, :symbolize_names => true)
+        end
+
+        def deliver args, headers={}
+          args = Promoted::Ruby::Client::Util.translate_args(args)
+
+          delivery_request_builder = RequestBuilder.new({
+            only_log: @only_log,
+            should_apply_treatment: @should_apply_treatment
+          })
+          delivery_request_builder.set_request_params(args)
+
+          if perform_checks?
+            Promoted::Ruby::Client::Settings.check_that_log_ids_not_set!(args)
+          end
+
+          pre_delivery_fillin_fields log_request_builder
+  
+          response_insertions = []
+          cohort_membership_to_log = nil
+          insertions_from_promoted = false
+          if !delivery_request_builder.only_log
+            cohort_membership_to_log = delivery_request_builder.new_cohort_membership_to_log
+          end
+  
+          if delivery_request_builder.should_apply_treatment
+            delivery_request_params = delivery_request_builder.delivery_request_params
+            response = send_request(delivery_request_params, @delivery_endpoint, @delivery_timeout_millis, headers)
+            insertions_from_promoted = true;
+            response_insertions = delivery_request_builder.fill_details_from_response(response.insertion)
+          end
+  
+          request_to_log = nil
+          if !insertions_from_promoted
+            request_to_log = delivery_request_builder.request
+            size = delivery_request_builder.request.dig(:paging, :size)
+            response_insertions = size.present? ? delivery_request_builder.full_insertion[0..size] : delivery_request_builder.full_insertion
+          end
+
+          if request_to_log
+            request_to_log[:request_id] = SecureRandom.uuid if not request_to_log[:request_id]
+            add_missing_ids_on_insertions! request_to_log, response_insertions
+          end
+
+          log_request_builder = RequestBuilder.new({
```

Comment:
@danbosnichill I went back and attempted to finish this, it's a little tricky, would you mind taking a look and letting me know if it is moving in a plausible direction, and if so I will attempt to pound it with tests in a follow-up PR? Thanks!